osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/threads/RecoverKeyProcess.java

```diff
@@ -75,25 +82,37 @@ public void run() {
 
         LOGGER.info("Attempting key recovery for device {}", this.deviceIdentification);
 
+        //FIXME why try and connect to device if you don't do anything with the result?!?
         try {
-            this.initDevice();
+            this.findDevice();
         } catch (final Exception e) {
-            LOGGER.error("Unexpected exception: {}", e);
+            LOGGER.error("Unexpected exception", e);
         }
-        if (!this.device.hasNewSecurityKey()) {
+
+        if (!this.securityKeyService.hasNewSecretOfType(this.deviceIdentification, E_METER_AUTHENTICATION)) {
+            //TODO: log something or throw error/exception?
             return;
         }
 
-        if (this.canConnect()) {
-            this.promoteInvalidKey();
+        if (this.canConnectUsingNewKeys()) {
+            try {
+                this.securityKeyService.activateNewKey(this.deviceIdentification, E_METER_ENCRYPTION);
+            } catch (ProtocolAdapterException e) {
+                throw new RecoverKeyException(e.getMessage(), e);
+            }
+            try {
+                this.securityKeyService.activateNewKey(this.deviceIdentification, E_METER_AUTHENTICATION);
+            } catch (ProtocolAdapterException e) {
+                throw new RecoverKeyException(e.getMessage(), e);
+            }
         }
+        //FIXME else: throw recovery error?
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1134)